File: python/examples/03_resume_thread.py
from __future__ import annotations

import asyncio

from codex_harness_kit import ApprovalPolicy, Session

from _common import CLIENT_INFO, print_section, require_codex_cli


async def main() -> None:
    require_codex_cli()
    print_section("Resume Thread Across Sessions")

    async with await Session.create(
        client_info=CLIENT_INFO,
        approval_policy=ApprovalPolicy.auto_accept(),
    ) as first_session:
        thread = await first_session.start_thread({"ephemeral": False})
        first_answer = await thread.ask("Reply with exactly FIRST.")
        thread_id = thread.id

        print(f"Created persistent thread: {thread_id}")
        print(f"First answer: {first_answer.strip()}")

    print("Closed the first session. Reconnecting with a fresh session.")

    async with await Session.create(
        client_info=CLIENT_INFO,
        approval_policy=ApprovalPolicy.auto_accept(),
    ) as second_session:
        thread = await second_session.resume_thread(thread_id)
        snapshot = await second_session.read_thread(thread.id, include_turns=True)
        second_answer = await thread.ask("Reply with exactly SECOND.")

        print(f"Resumed thread: {thread.id}")
        print(f"Turns already on thread: {len(snapshot.get('turns', []))}")
        print(f"Second answer: {second_answer.strip()}")


if __name__ == "__main__":
    asyncio.run(main())

File: typescript/examples/03_resume_thread.ts
import {
  ApprovalPolicy,
  ProtocolClientError,
  Session,
  TurnFailedError,
  threadParams,
} from "../src/index.js";
import { CLIENT_INFO, printSection, withRetry } from "./_common.js";

const firstSession = await Session.create({
  approvalPolicy: ApprovalPolicy.autoAccept(),
  clientInfo: CLIENT_INFO,
});

let threadId = "";
let initialTurnCompleted = false;
printSection("Resume Thread Across Sessions");
try {
  const thread = await firstSession.startThread(threadParams({ ephemeral: false }));
  threadId = thread.id;
  console.log(`Created persistent thread: ${thread.id}`);
  try {
    const answer = await withRetry("initial ask()", async () => await thread.ask("Reply with exactly FIRST."));
    console.log(`First answer: ${answer.trim()}`);
    initialTurnCompleted = true;
  } catch (error) {
    if (error instanceof TurnFailedError) {
      console.log(`Initial turn failed: ${error.message}`);
    }
    if (!(error instanceof TurnFailedError)) {
      throw error;
    }
  }
} finally {
  await firstSession.close();
}

if (!initialTurnCompleted) {
  console.log("Skipping resume because the initial turn did not complete successfully.");
} else {
console.log("Closed the first session. Reconnecting with a fresh session.");

const secondSession = await Session.create({
  approvalPolicy: ApprovalPolicy.autoAccept(),
  clientInfo: CLIENT_INFO,
});

try {
  try {
    const resumed = await secondSession.resumeThread(threadId);
    const snapshot = await secondSession.readThread(resumed.id, true);
    console.log(`Turns already on thread: ${snapshot.turns.length}`);

    try {
      const answer = await withRetry("resumed ask()", async () => await resumed.ask("Reply with exactly SECOND."));
      console.log(`Second answer: ${answer.trim()}`);
    } catch (error) {
      if (error instanceof TurnFailedError) {
        console.log(`Resumed turn failed: ${error.message}`);
      } else {
        throw error;
      }
    }
  } catch (error) {
    if (error instanceof ProtocolClientError && error.code === -32600) {
      console.log(`Resume is not available on this local codex app-server instance: ${error.message}`);
    } else {
      throw error;
    }
  }
} finally {
  await secondSession.close();
}
}
